litt/cli/prune.go

```diff
@@ -0,0 +1,230 @@
+}
+
+// prune deletes data from a littDB database/snapshot.
+func prune(logger logging.Logger, sources []string, allowedTables []string, maxAgeSeconds uint64, fsync bool) error {
+	allowedTablesSet := make(map[string]struct{})
+	for _, table := range allowedTables {
+		allowedTablesSet[table] = struct{}{}
+	}
+
+	// Forbid touching tables in active use.
+	releaseLocks, err := util.LockDirectories(logger, sources, util.LockfileName, fsync)
+	if err != nil {
+		return fmt.Errorf("failed to acquire locks on paths %v: %w", sources, err)
+	}
+	defer releaseLocks()
+
+	// Determine which tables to prune.
+	var tables []string
+	foundTables, err := lsPaths(logger, sources, false, fsync)
+	if err != nil {
+		return fmt.Errorf("failed to list tables in paths %v: %w", sources, err)
+	}
+	if len(allowedTables) == 0 {
+		tables = foundTables
+	} else {
+		for _, table := range foundTables {
+			if _, ok := allowedTablesSet[table]; ok {
+				tables = append(tables, table)
+			}
+		}
+	}
+
+	// Prune each table.
+	for _, table := range tables {
+		bytesDeleted, err := pruneTable(logger, sources, table, maxAgeSeconds, fsync)
+		if err != nil {
+			return fmt.Errorf("failed to prune table %s in paths %v: %w", table, sources, err)
+		}
+
+		logger.Infof("Deleted %s from table '%s'.", common.PrettyPrintBytes(bytesDeleted), table)
+	}
+
+	return nil
+}
+
+// pruneTable performs offline garbage collection on a LittDB database/snapshot.
+func pruneTable(
+	logger logging.Logger,
+	sources []string,
+	tableName string,
+	maxAgeSeconds uint64,
+	fsync bool) (uint64, error) {
+
+	errorMonitor := util.NewErrorMonitor(context.Background(), logger, nil)
+
+	segmentPaths, err := segment.BuildSegmentPaths(sources, "", tableName)
+	if err != nil {
+		return 0, fmt.Errorf("failed to build segment paths for table %s at paths %v: %w",
+			tableName, sources, err)
+	}
+
+	lowestSegmentIndex, highestSegmentIndex, segments, err := segment.GatherSegmentFiles(
+		logger,
+		errorMonitor,
+		segmentPaths,
+		false,
+		time.Now(),
+		true,
+		fsync)
+	if err != nil {
+		return 0, fmt.Errorf("failed to gather segment files for table %s at paths %v: %w",
+			tableName, sources, err)
+	}
+
+	if len(segments) == 0 {
+		return 0, fmt.Errorf("no segments found for table %s at paths %v", tableName, sources)
+	}
+
+	// Determine if we are working on the snapshot directory (i.e. the directory with symlinks to the segments).
+	isSnapshot, err := segments[lowestSegmentIndex].IsSnapshot()
+	if err != nil {
+		return 0, fmt.Errorf("failed to check if segment %d is a snapshot: %w", lowestSegmentIndex, err)
+	}
+
+	if isSnapshot {
+		// If we are dealing with a snapshot, respect the snapshot upper bound specified by LittDB.
+		if len(sources) > 1 {
+			return 0, fmt.Errorf("this is a symlinked snapshot directory, " +
+				"snapshot directory cannot be spread across multiple sources")
+		}
+		upperBoundFile, err := disktable.LoadBoundaryFile(false, path.Join(sources[0], tableName))
```

Comment:
same question here, why pass in `false`, instead of the defined enum type?

Reply:
Similar to before. I think you make a valid point, but I don't want to change it in this PR due to the merge conflicts it would cause.